hphp/runtime/vm/jit/vasm-ppc64.cpp

```diff
@@ -1564,6 +1438,142 @@ void lowerForPPC64(Vunit& unit) {
+  enum class AddressModes {
+    Invalid        = 0,
+    Disp           = 1, // Displacement
+    Base           = 2, // Base
+    BaseDisp       = 3, // Base+Displacement
+    Index          = 4, // Index
+    IndexDisp      = 5, // Index+Dispacement
+    IndexBase      = 6, // Index+Base
+    IndexBaseDisp  = 7  // Index+Base+Displacement
+  };
+
+  AddressModes mode = static_cast<AddressModes>(
+                    (((p.disp != 0)     & 0x1) << 0) |
+                    ((p.base.isValid()  & 0x1) << 1) |
+                    ((p.index.isValid() & 0x1) << 2));
+
+  // Index can never be used directly if shifting is necessary. Handling it here
+  uint8_t shift = p.scale == 2 ? 1 :
+                  p.scale == 4 ? 2 :
+                  p.scale == 8 ? 3 : 0;
+
+  if (p.index.isValid() && shift) {
+    Vreg shifted_index_reg = v.makeReg();
+    v << shlqi{shift, p.index, shifted_index_reg, VregSF(RegSF{0})};
+    p.index = shifted_index_reg;
+    p.scale = 1;  // scale is now normalized.
+  }
+
+  // taking care of the displacement, in case it is > 16bits
+  Vreg disp_reg;
+  bool patched_disp = patchImm(Immed(p.disp), v, disp_reg);
+  switch (mode) {
+    case AddressModes::Base:
+    case AddressModes::IndexBase:
+      // ppc64 can handle these address modes. Nothing to do here.
+      break;
+
+    case AddressModes::BaseDisp:
+      // ppc64 can handle this address mode if displacement < 16bits
+      if (patched_disp) {
+        // disp is loaded on a register. Change address mode to kBase_Index
+        p.index = disp_reg;
+        p.disp = 0;
+      }
+      break;
+
+    case AddressModes::Index:
+        // treat it as kBase to avoid a kIndex_Disp asm handling.
+        std::swap(p.base, p.index);
+      break;
+
+    case AddressModes::Disp:
+    case AddressModes::IndexDisp:
+      if (patched_disp) {
+        // disp is loaded on a register. Change address mode to kBase_Index
+        p.base = disp_reg;
+        p.disp = 0;
+      }
+      else {
+        // treat it as kBase_Disp to avoid a kIndex_Disp asm handling.
+        p.base = p.index;
+      }
+      break;
+
+    case AddressModes::IndexBaseDisp: {
+      // This mode is not supported: Displacement will be embedded on Index
+      Vreg index_disp_reg = v.makeReg();
+      lowerImm(Immed(p.disp), v, disp_reg);
+      v << addq{disp_reg, p.index, index_disp_reg, VregSF(RegSF{0})};
+      p.index = index_disp_reg;
+      p.disp = 0;
+      break;
+    }
+
+    case AddressModes::Invalid:
+      assert(false && "Invalid address mode");
+      break;
+  }
+}
+
+/*
+ * Every vasm that has a Vptr parameter must be handled
+ */
+struct VptrFixVisit {
+explicit VptrFixVisit(Vout& v) : m_v(v) {}
+  Vout& m_v;
+
+  template<class T> void def(T) {}
+  template<class T> void imm(const T&) {}
+  template<class T> void across(T) {}
+  template<class T, class H> void defHint(T, H) {}
+  template<class T, class H> void useHint(T, H) {}
```

Comment:
I'm pretty sure there are not use-hints with Vptr operands, but it would be better to have

```
template<class T, class H> void useHint(T r, H) { use(r); }
```

just in case someday we decide to go wild.